### PR TITLE
Minor optimization of inftrees.c based on profiling

### PR DESCRIPTION
**Pull request by Dead2**
Reorder an if chain to make the most likely branch come first, and provide a compiler hint that two branches are very unlikely.
Results in about 0.25% speedup in my tests.

### CODE REVIEW

**Comment by Dead2** [comment]
### Baseline e56238cbc4433b71b75f08deac35addb8426fef6 x86-64
```
   text    data     bss     dec     hex filename
 113433    1376      32  114841   1c099 libz-ng.so.2

 Tool: minigzip   Levels: 1-9
 Runs: 30         Trim worst: 15

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.165%      2.975/3.007/3.022/0.014        1.647/1.675/1.690/0.015      264,231,515
 2     43.875%      2.950/2.991/3.006/0.013        0.992/1.008/1.022/0.011      124,277,644
 3     42.382%      2.980/2.997/3.010/0.008        0.800/0.813/0.823/0.007      100,039,412
 4     41.669%      2.730/2.749/2.766/0.012        0.616/0.631/0.643/0.007       78,685,354
 5     41.215%      2.924/2.933/2.946/0.007        0.565/0.580/0.589/0.007       71,342,944
 6     41.036%      2.804/2.823/2.834/0.009        0.461/0.473/0.478/0.005       58,118,026
 7     40.781%      2.723/2.738/2.748/0.008        0.352/0.362/0.371/0.006       44,921,649
 8     40.702%      3.009/3.015/3.021/0.004        0.310/0.317/0.324/0.005       38,429,965
 9     40.698%      2.774/2.793/2.801/0.009        0.253/0.264/0.270/0.005       32,021,745

 avg1  42.947%                        2.894                          0.680
 tot                                390.693                         91.864      812,068,254
 
  Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.165%      2.979/3.008/3.023/0.013        1.645/1.675/1.688/0.012      264,231,515
 2     43.875%      2.969/3.004/3.021/0.014        0.986/1.008/1.019/0.009      124,277,644
 3     42.382%      2.996/3.020/3.033/0.010        0.790/0.813/0.824/0.009      100,039,412
 4     41.669%      2.720/2.737/2.746/0.008        0.610/0.639/0.646/0.010       78,685,354
 5     41.215%      2.954/2.977/2.987/0.009        0.559/0.581/0.591/0.009       71,342,944
 6     41.036%      2.781/2.817/2.829/0.012        0.438/0.470/0.478/0.010       58,118,026
 7     40.781%      2.720/2.741/2.750/0.008        0.348/0.368/0.378/0.008       44,921,649
 8     40.702%      2.994/3.007/3.016/0.007        0.301/0.317/0.324/0.006       38,429,965
 9     40.698%      2.783/2.798/2.805/0.006        0.253/0.259/0.263/0.003       32,021,745

 avg1  42.947%                        2.901                          0.681
 tot                                391.637                         91.943      812,068,254
```

### PR1006 841d7b918407a4496cbae59161628d3bfda572f4 x86-64
```
   text    data     bss     dec     hex filename
 113433    1376      32  114841   1c099 libz-ng.so.2

 Tool: minigzip   Levels: 1-9
 Runs: 30         Trim worst: 15

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.165%      2.969/3.000/3.016/0.014        1.644/1.668/1.689/0.017      264,231,515
 2     43.875%      2.964/2.980/2.992/0.008        0.996/1.016/1.026/0.009      124,277,644
 3     42.382%      2.972/2.992/3.004/0.011        0.788/0.810/0.818/0.008      100,039,412
 4     41.669%      2.717/2.739/2.751/0.010        0.621/0.632/0.645/0.009       78,685,354
 5     41.215%      2.908/2.925/2.936/0.008        0.547/0.576/0.587/0.010       71,342,944
 6     41.036%      2.816/2.827/2.837/0.007        0.463/0.472/0.480/0.004       58,118,026
 7     40.781%      2.703/2.726/2.737/0.009        0.347/0.363/0.370/0.007       44,921,649
 8     40.702%      2.992/3.010/3.020/0.009        0.296/0.311/0.319/0.006       38,429,965
 9     40.698%      2.782/2.791/2.796/0.004        0.249/0.261/0.266/0.004       32,021,745

 avg1  42.947%                        2.888                          0.679
 tot                                389.859                         91.644      812,068,254

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.165%      2.935/2.991/3.006/0.018        1.646/1.669/1.685/0.011      264,231,515
 2     43.875%      2.983/3.002/3.011/0.009        0.986/1.005/1.016/0.009      124,277,644
 3     42.382%      2.999/3.014/3.028/0.010        0.801/0.818/0.825/0.007      100,039,412
 4     41.669%      2.716/2.731/2.741/0.008        0.601/0.630/0.641/0.011       78,685,354
 5     41.215%      2.927/2.964/2.978/0.014        0.567/0.581/0.590/0.008       71,342,944
 6     41.036%      2.791/2.807/2.819/0.008        0.453/0.467/0.477/0.007       58,118,026
 7     40.781%      2.708/2.731/2.742/0.009        0.353/0.367/0.374/0.006       44,921,649
 8     40.702%      2.990/3.003/3.011/0.006        0.303/0.312/0.316/0.004       38,429,965
 9     40.698%      2.774/2.786/2.795/0.007        0.249/0.261/0.269/0.005       32,021,745

 avg1  42.947%                        2.892                          0.679
 tot                                390.427                         91.646      812,068,254
```

**Comment by Dead2** [comment]
### PR1006 9dd34ecb47ba860c6db6f9cd817270641f77e6ec x86-64
```
    text    data     bss     dec     hex filename
 113433    1376      32  114841   1c099 libz-ng.so.2

 Tool: minigzip   Levels: 1-9
 Runs: 30         Trim worst: 15

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.165%      2.975/3.011/3.028/0.015        1.641/1.673/1.688/0.012      264,231,515
 2     43.875%      2.963/2.994/3.008/0.014        0.993/1.005/1.016/0.008      124,277,644
 3     42.382%      2.986/3.014/3.024/0.010        0.781/0.812/0.825/0.011      100,039,412
 4     41.669%      2.717/2.730/2.740/0.008        0.608/0.635/0.648/0.013       78,685,354
 5     41.215%      2.939/2.967/2.977/0.010        0.557/0.580/0.590/0.009       71,342,944
 6     41.036%      2.790/2.812/2.824/0.010        0.460/0.472/0.480/0.006       58,118,026
 7     40.781%      2.711/2.729/2.741/0.011        0.340/0.362/0.370/0.008       44,921,649
 8     40.702%      2.992/3.006/3.012/0.006        0.299/0.312/0.320/0.006       38,429,965
 9     40.698%      2.772/2.789/2.797/0.007        0.249/0.258/0.262/0.003       32,021,745

 avg1  42.947%                        2.895                          0.679
 tot                                390.784                         91.633      812,068,254
 ```

**Review by nmoinvaz — APPROVED** [review]

**Comment by mtl1979** [comment]
0.25% is well within error margin.

**Comment by Dead2** [comment]
@mtl1979 Are you saying you don't think there is an actual performance improvement with this PR?
It seems very consistent here, both in benchmarking and profiling, feel free to perform your own profiling to see how often these are used in workloads you care about.

If you think the improvement is just small, well there is not much to be done about that, this is a simple and minor optimization after all.

**Comment by mtl1979** [comment]
@Dead2 I'm not saying that there can't be improvement, I'm just not trusting the number... Most results from benchmarks I have seen are pretty inconsistent as they use only one type of file with predictable contents. Even if you run 70 iterations of test with one file, the error margin is more than if you run 10 or 20 iterations of 7 files. Concatenating multiple files to a larger file is not same as running test with same files non-concatenated.

Basically doing same operation over and over will just let the operating system and processor learn from the data and cache enough data that skew the results.

**Comment by Dead2** [comment]
@mtl1979 Feel free to contribute improvements to benchmarking tools

**Comment by Dead2** [comment]
### Baseline 834e7d83eae9ff7bfb1b9769d9f311ea7ca8e746 aarch64
```
   text    data     bss     dec     hex filename
 105194    1576      16  106786   1a122 libz-ng.so.2

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      2.746/2.764/2.780/0.011        0.965/0.993/1.015/0.013       42,618,858
 2     43.879%      2.819/2.851/2.870/0.017        0.563/0.601/0.631/0.022       20,714,938
 3     42.383%      3.713/3.758/3.780/0.020        0.565/0.580/0.592/0.008       20,008,755
 4     41.660%      2.768/2.802/2.818/0.015        0.353/0.372/0.381/0.008       13,111,391
 5     41.216%      3.014/3.031/3.051/0.012        0.365/0.375/0.383/0.005       12,971,746
 6     41.039%      3.505/3.539/3.555/0.014        0.358/0.372/0.384/0.009       12,916,025
 7     40.778%      2.239/2.270/2.284/0.011        0.168/0.175/0.181/0.004        6,416,924
 8     40.704%      2.769/2.800/2.809/0.010        0.168/0.179/0.188/0.007        6,405,244
 9     40.696%      2.983/3.016/3.029/0.013        0.160/0.172/0.180/0.007        6,404,084

 avg1  42.947%                        2.981                          0.424
 tot                                402.450                         57.285      141,567,965
```

### PR #1006 9dd34ecb47ba860c6db6f9cd817270641f77e6ec aarch64
```
   text    data     bss     dec     hex filename
 105234    1576      16  106826   1a14a libz-ng.so.2

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      2.741/2.775/2.802/0.018        0.947/0.989/1.000/0.015       42,618,858
 2     43.879%      2.785/2.838/2.865/0.024        0.561/0.595/0.611/0.014       20,714,938
 3     42.383%      3.698/3.746/3.779/0.022        0.550/0.577/0.595/0.014       20,008,755
 4     41.660%      2.765/2.796/2.815/0.016        0.365/0.377/0.384/0.006       13,111,391
 5     41.216%      3.006/3.029/3.046/0.014        0.357/0.370/0.382/0.008       12,971,746
 6     41.039%      3.507/3.531/3.540/0.009        0.340/0.363/0.376/0.011       12,916,025
 7     40.778%      2.239/2.268/2.281/0.013        0.164/0.177/0.184/0.007        6,416,924
 8     40.704%      2.789/2.803/2.808/0.005        0.160/0.176/0.184/0.008        6,405,244
 9     40.696%      2.994/3.017/3.029/0.010        0.156/0.172/0.180/0.007        6,404,084

 avg1  42.947%                        2.978                          0.422
 tot                                402.027                         56.930      141,567,965
```

Compiled code size increased by 40 bytes, likely due to moving the two unlikely blocks of code out of the hot-path of the function.
0.62% faster inflate.
